compiler/src/symboltable.rs

```diff
@@ -747,10 +747,13 @@ impl SymbolTableBuilder {
             // Role already set..
             match role {
                 SymbolUsage::Global => {
-                    return Err(SymbolTableError {
-                        error: format!("name '{}' is used prior to global declaration", name),
-                        location,
-                    })
+                    let symbol = table.symbols.get(name).unwrap();
+                    if symbol.is_assigned {
```

Comment:
How about this case:
```python
def foo():
    print(x)
    global x
```

In this case `x` is not assigned, but global is not allowed anyways.

Instead of `symbol.is_assigned`, check if `symbol.scope` is set to `Global` before.

Reply:
I think it is important that the variable is used or referenced in same scope rather than global declaration of variable, so I used `symbol.is_assigned` and `symbol.is_referenced`. Is my solution wrong?

Reply:
I think your solution is wrong, and the only legal re-declaration of a global is when the variable was previously marked as global, so in that case, check for `symbol.is_global`, this should be true, otherwise we face an error.

Reply:
Can I get a review again?